ledger/catchupaccessor.go

```diff
@@ -391,34 +391,184 @@ func (c *CatchpointCatchupAccessorImpl) processStagingBalances(ctx context.Conte
 	return err
 }
 
-// EvictAsNeeded calls Evict on the cachedTrie periodically, or once we're done updating the trie.
-func (progress *CatchpointCatchupAccessorProgress) EvictAsNeeded(balancesCount uint64) (err error) {
-	if progress.cachedTrie == nil {
-		return nil
+// BuildMerkleTrie would process the catchpointpendinghashes and insert all the items in it into the merkle trie
+func (c *CatchpointCatchupAccessorImpl) BuildMerkleTrie(ctx context.Context, progressUpdates func(uint64)) (err error) {
+	wdb := c.ledger.trackerDB().wdb
+	rdb := c.ledger.trackerDB().rdb
+	err = wdb.Atomic(func(ctx context.Context, tx *sql.Tx) (err error) {
+		// creating the index can take a while, so ensure we don't generate false alerts for no good reason.
+		db.ResetTransactionWarnDeadline(ctx, tx, time.Now().Add(120*time.Second))
+		return createCatchpointStagingHashesIndex(ctx, tx)
+	})
+	if err != nil {
+		return
 	}
-	if progress.evictFrequency == 0 {
-		progress.evictFrequency = trieRebuildCommitFrequency
-	}
-	// periodically, perform commit & evict to flush it to the disk and rebalance the cache memory utilization.
-	if (progress.ProcessedAccounts/progress.evictFrequency) < ((progress.ProcessedAccounts+balancesCount)/progress.evictFrequency) ||
-		(progress.ProcessedAccounts+balancesCount) == progress.TotalAccounts {
-		evictStart := time.Now()
-		_, err = progress.cachedTrie.Evict(true)
-		if err == nil {
-			evictDelta := time.Now().Sub(evictStart)
-			if evictDelta > maxMerkleTrieEvictionDuration {
-				if progress.evictFrequency > minMerkleTrieEvictFrequency {
-					progress.evictFrequency /= 2
+
+	wg := sync.WaitGroup{}
+	wg.Add(2)
+	errChan := make(chan error, 2)
+
+	writerQueue := make(chan [][]byte, 16)
+	c.ledger.setSynchronousMode(ctx, c.ledger.accountsRebuildSynchronousMode)
+	defer c.ledger.setSynchronousMode(ctx, c.ledger.synchronousMode)
+
+	// starts the hashes reader
+	go func() {
+		defer wg.Done()
+		defer close(writerQueue)
+
+		err := rdb.Atomic(func(transactionCtx context.Context, tx *sql.Tx) (err error) {
+			it := makeCatchpointPendingHashesIterator(trieRebuildAccountChunkSize, tx)
+			var hashes [][]byte
+			for {
+				hashes, err = it.Next(transactionCtx)
+				if err != nil {
+					break
 				}
-			} else {
-				progress.evictFrequency *= 2
-				if progress.evictFrequency > trieRebuildCommitFrequency {
-					progress.evictFrequency = trieRebuildCommitFrequency
+				if len(hashes) > 0 {
+					writerQueue <- hashes
+				}
+				if len(hashes) != trieRebuildAccountChunkSize {
```

Comment:
This will be equivalent to:
len(hashes) < trieRebuildAccountChunkSize
right?

Reply:
correct. The `it.Next` is expected to return a `trieRebuildAccountChunkSize` sized arrays unless it's the last iteration.